x-pack/solutions/observability/plugins/investigate_app/scripts/evaluate/kibana_client.ts

```diff
@@ -0,0 +1,326 @@
+        );
+      }
+      const from = datemath.parse(fromOffset, { forceNow: new Date(alertStart) })?.valueOf()!;
+      const to = datemath.parse(toOffset, { forceNow: new Date(alertStart) })?.valueOf()!;
+      return {
+        from,
+        to,
+      };
+    }
+
+    async function createInvestigation({
+      alertId,
+      from,
+      to,
+    }: {
+      alertId: string;
+      from: number;
+      to: number;
+    }): Promise<string> {
+      const body = {
+        id: uuidv4(),
+        title: 'Investigate Custom threshold breached',
+        params: {
+          timeRange: {
+            from,
+            to,
+          },
+        },
+        tags: [],
+        origin: {
+          type: 'alert',
+          id: alertId,
+        },
+        externalIncidentUrl: null,
+      };
+
+      const response = await that.axios.post(
+        that.getUrl({
+          pathname: '/api/observability/investigations',
+        }),
+        body
+      );
+      return response.data.id;
+    }
+
+    async function deleteInvestigation({
+      investigationId,
+    }: Parameters<RCAChatClient['deleteInvestigation']>[0]) {
+      await that.axios.delete(
+        that.getUrl({
+          pathname: `/api/observability/investigations/${investigationId}`,
+        })
+      );
+    }
+
+    async function rootCauseAnalysis({
+      connectorIdOverride,
+      investigationId,
+      from,
+      to,
+      alert,
+    }: {
+      connectorIdOverride?: string;
+      investigationId: string;
+      from: string;
+      to: string;
+      alert?: EcsFieldsResponse;
+    }) {
+      const chat$ = defer(() => {
+        that.log.debug(`Calling root cause analysis API`);
+        const serviceName = alert?.['service.name'] as string | undefined;
+        if (!alert) {
+          throw new Error(
+            'Alert not found. Please ensure you have loaded test fixture data prior to running tests.'
+          );
+        }
+        if (!serviceName) {
+          throw new Error(
+            'Service name is missing from the alert data. Please double check your alert fixture.'
+          );
+        }
+        const context = getRCAContext(alert, serviceName);
+        const body = {
+          investigationId,
+          connectorId: connectorIdOverride || '34002e09-65be-4332-9084-2649c86abacb',
+          context,
+          rangeFrom: from,
+          rangeTo: to,
+          serviceName: 'controller',
+          completeInBackground: false,
+        };
+
+        return that.axios.post(
+          that.getUrl({
+            pathname: '/internal/observability/investigation/root_cause_analysis',
+          }),
+          body,
+          { responseType: 'stream', timeout: NaN }
+        );
+      }).pipe(
+        switchMap((response) => streamIntoObservable(response.data)),
+        serializeRCAResponse(),
+        tap((event) => {
+          switch (true) {
+            case event.includes('"name":"endProcessAndWriteReport"'):
```

Comment:
why not parse the JSON first?

Reply:
Because at this point in time, the stream JSON object might be incomplete and therefore would be unable to parse. Large JSON objects from investigate tool call typically flow in in two chunks.

Reply:
If you use the tools from `@kbn/sse-utils-client` that should clean this up. These are server-sent events so they adhere to a specification and there are parsers for it.